skbio/sequence/_interval.py

```diff
@@ -0,0 +1,121 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+from __future__ import absolute_import, division, print_function
+
+from ._feature import Feature
+from .intersection import Interval, IntervalTree
+import itertools
+from collections import defaultdict
+
+class IntervalMetadata():
+    ''' Store metadata corresponding to Features and Intervals.
+
+    There are two underlying data structures namely
+    1. An IntervalTree (interval_tree)
+    2. A bidirectional hash table (feature_table)
+
+    Attributes
+    ----------
+    interval_tree
+    feature_table
+
+    '''
+    def __init__(self):
+        self.interval_tree = IntervalTree()
+        # maps features to intervals and vice versa
+        self.feature_table = defaultdict(list)
+
+    def add(self, x, kw='location'):
+        """ Adds a feature to the metadata object.
+
+        Parameters
+        ----------
+        x : skbio.sequence.feature
+            The feature object being added.
+        kw : str
+            The keyword associated with the intervals.
+
+        Note
+        ----
+        The intervals associated with a feature is assumed to be under
+        the `location` keyword by default.
+        """
+        if not isinstance(x, Feature):
+            raise ValueError('x is not an instance of `Feature`')
+
+        intervals = x[kw]
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        for interval in intervals:
+            self.interval_tree.add(interval[0], interval[1], x)
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        for k, v in x.items():
+            if x not in self.feature_table:
+                self.feature_table[(k, v)] = [x]
+            else:
+                self.feature_table[(k, v)].append(x)
+
+
+    def query(self, *args, **kwargs):
+        """
+        Parameters
+        ----------
+        kwargs : dict
+            Keyword arguments of feature name and feature value, which can
+            be passed to ``dict``.  This is used to specify the search parameters.
+            If the `location` keyword is passed, then an interval lookup will be
+            performed.
```

Comment:
I think making a _query_interval method is fine.  However, making a private
_query_feature is a bit redundant.  That method is already super simple.

On Fri, Feb 5, 2016 at 2:11 PM, Zech Xu <notifications@github.com> wrote:

> In skbio/sequence/_interval.py
> <https://github.com/RNAer/scikit-bio/pull/1#discussion_r52079769>:
>
> > +        for k, v in x.items():
> > +            if x not in self.feature_table:
> > +                self.feature_table[(k, v)] = [x]
> > +            else:
> > +                self.feature_table[(k, v)].append(x)
> > +
> > +
> > +    def query(self, *args, **kwargs):
> > +        """
> > +        Parameters
> > +        ----------
> > +        kwargs : dict
> > +            Keyword arguments of feature name and feature value, which can
> > +            be passed to ``dict``.  This is used to specify the search parameters.
> > +            If the `location` keyword is passed, then an interval lookup will be
> > +            performed.
>
> a pin for updating doc. what do you think of this func wrapping on 2
> private funcs doing real heavy lifting? I imagine you will need query for
> location and query for attributes for [ ] overloading.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RNAer/scikit-bio/pull/1/files#r52079769>.
>